agent/eni/iphelperwrapper/iphelper_windows.go

```diff
@@ -0,0 +1,163 @@
+	// Also, note that the initial notification will only be sent once. Therefore, this delay would be for the first communication only
+	DelayInCallbackMessageForInitialNotification = 10 * time.Millisecond
+)
+
+var (
+	moduleiphlpapi = windows.NewLazySystemDLL("iphlpapi.dll")
+
+	procNotifyIPInterfaceChange = moduleiphlpapi.NewProc("NotifyIpInterfaceChange")
+	proccancelMibChangeNotify2  = moduleiphlpapi.NewProc("CancelMibChangeNotify2")
+
+	// Added these to allow unit testing for this module. We will swap these System calls with dummy calls in unit tests
+	funcNotifyIPInterfaceChange = procNotifyIPInterfaceChange.Call
+	funcCancelMibChangeNotify2  = proccancelMibChangeNotify2.Call
+
+	// The notification channel is used for communication between the notification callback from Windows APIs and the consumer
+	notificationChannel chan int
+)
+
+// Interface for using the monitor
+// We will mock this interface while unit testing other packages
+type InterfaceMonitor interface {
+	StartMonitor(notification chan int) error
+	Close() error
+}
+
+type monitoringInterface struct {
+	notificationHandle windows.Handle
+}
+
+// Returns a new monitoringInterface which conforms to InterfaceMonitor interface
+func NewMonitor() (monitor *monitoringInterface) {
+	monitor = &monitoringInterface{}
+
+	notificationHandle := windows.Handle(0)
+	monitor.notificationHandle = notificationHandle
+
+	return
+}
+
+// This method is used to start the monitoring of new interfaces attached to the instance.
+// We register a notification callback with Windows API for any changes in instance interfaces (NotifyIPInterfaceChange).
+// Initial notification received within timeout is considered as a confirmation of successful registration
+// notificationChannel is used for any communication between the callback and the consumer
+func (monitor *monitoringInterface) StartMonitor(notification chan int) error {
+	notificationChannel = notification
+
+	log.Info("Starting the monitor")
+	retVal, _, _ := funcNotifyIPInterfaceChange(syscall.AF_UNSPEC,
+		syscall.NewCallback(callback),
+		uintptr(unsafe.Pointer(&(monitor))),
+		1,
+		uintptr(unsafe.Pointer(&(monitor.notificationHandle))))
+
+	// If there is any error while registering callback then retVal will not be 0. Therefore, we will return with an error.
+	if retVal != 0 {
+		return errors.Errorf("error occured while calling Windows API : Error Msg : %s", syscall.Errno(retVal))
+	}
+
+	// Initial notification is sent immediately once the callback is registered.
+	// If the same is not received within a defined timeout then we assume an error and close the monitor
+initialNotificationLoop:
+	for {
+		select {
+		case event := <-notificationChannel:
+			if event == InitialNotification {
+				break initialNotificationLoop
+			}
+		case <-time.After(InitialNotificationTimeout):
+			log.Errorf("IPHelper : Initial notification not received within timeout. Closing the Monitor.")
```

Comment:
Please follow https://github.com/golang/go/wiki/CodeReviewComments for logging

Reply:
I have followed the existing logging format as used in Linux.
There aren't log message guidelines in the document.
Did you have anything specific in mind?